### PR TITLE
Auto-select edited text object when switching from text- to select-tool

### PR DESCRIPTION
**Pull request by forgottosave**
implements [#1169](https://github.com/xournalpp/xournalpp/issues/1169)

# Issue:
A requested enhancement was auto-selecting edited text objects after switching to a selection tool (TOOL_SELECT_RECT, TOOL_SELECT_REGION, TOOL_SELECT_OBJECT), so that you don't have to re-select a currently edited text object when wanting to move it.
GitHub Issue: [Switching to a selection tool should auto-select any currently editing text object #1169](https://github.com/xournalpp/xournalpp/issues/1169),
part of [Future Major Milestone](https://github.com/xournalpp/xournalpp/milestone/7)

# Implementation:
## general idea
When switching to a selection tool a new method in src/core/control/Control.cpp is being called: `selectAndEndText()`
This method:
- checks if the old tool was the text tool
- gets a reference to the Text object in the current TextEditor in PageView
- calls the `endText()` method to close the TextEditor and add the Text object to the layer (this method existed already)
- adds the Text object to the (if necessary new) Selection in Xournal
## changes needed in other files
1. `getActveTool()` in ToolHandler & `getToolType()` in Tool, to be able to check for the last used Tool
2. add respective headers
3. make removing the text editor in `endText()` in PageView optional (by default removing it, as it was before)
4. make deleting the text editor a separate method (avoid duplicate code)

This is my first contribution to this project, I hope it fits all the requirements. If anything's off I'd be happy about a short feedback so that I can fix it and prevent future complications, but I think everything should be alright :)

### CODE REVIEW

File: src/core/control/ToolHandler.cpp
Comment (Technius) [inline src/core/control/ToolHandler.cpp:202]: This should return a pointer; if `activeTool` is null, this will segfault.

File: src/core/control/Control.cpp
Comment (Technius) [inline src/core/control/Control.cpp:1091]: This seems to duplicate existing functionality. I think the logic can be cleaned up as follows:

* First, check whether there is a text editor, and whether the text is currently non-empty. If not, then return. (There is nothing to be done here).
* Save the pointer to the text element.
* Call `clearSelectionEndText()` to close the text editor. Note that if the text element was actually added, it should be on some layer on some page (see `XojPageView::endText()` in `PageView.cpp`).
* If the pointer to the text element is not null, then set the current selection to the text element.

To further clean this up, we can add the type of the selection tool as a function parameter and then move the `selectTool()` logic in `actionPerformed()` into this method, so that all we need to do is call this method with the right tool type.

If you want to follow my suggestion here, I would also suggest renaming this method to something like `handleSelectToolAction`.

File: src/core/gui/PageView.cpp
Comment (Technius) [inline src/core/gui/PageView.cpp:204]: To me, this doesn't seem to make sense in isolation. If `removeTextEditor` is false, the text editor will be left in an invalid state, since the text element will no longer be in the document anymore.

**Review by Technius — CHANGES_REQUESTED** [review]
Thanks for the PR! I have a few major concerns.

Would also like to get a second opinion from @bhennion, who has refactored some of these files recently.

File: src/core/control/Control.cpp
Comment (forgottosave) [reply src/core/control/Control.cpp:1091]: I'm not sure if this would actually clean up code.

File: src/core/control/ToolHandler.cpp
Comment (forgottosave) [reply src/core/control/ToolHandler.cpp:202]: True

File: src/core/control/Control.cpp
Comment (forgottosave) [reply src/core/control/Control.cpp:1091]: Thank you very much. I think this really cleans up a lot and works just as fine.
Although I'm not sure about moving the `selectTool()` logic in `actionPerformed()` into this method. It only requires another check if the selected tool is one of the selection tools every time it is called.
I still implemented it, but would recommend checking on this again.

**Comment by forgottosave** [comment]
Looking at the error in [Build_Test_Stage Test Xournal++ on MacOS](https://dev.azure.com/xournalpp/xournalpp/_build/results?buildId=21713&view=logs&jobId=69082843-5ef9-5486-682b-24fa9b8c0790) I'm not sure where the issue is something I changed? Could it be an issue with the tests?

> Archive:  ninja-mac.zip
  End-of-central-directory signature not found.  Either this file is not
  a zipfile, or it constitutes one disk of a multi-part archive.  In the
  latter case the central directory and zipfile comment will be found on
  the last disk(s) of this archive.
unzip:  cannot find zipfile directory in one of ninja-mac.zip or
        ninja-mac.zip.zip, and cannot find ninja-mac.zip.ZIP, period.
##[error]Bash exited with code '9'.

File: src/core/control/Control.cpp
Comment (Technius) [inline src/core/control/Control.cpp:1782]: Based on the way this is implemented, I don't think this method needs to be renamed.

File: src/core/control/Control.cpp
Comment (Technius) [inline src/core/control/Control.cpp:1786]: nit: instead of directly checking `type == TOOL_SELECT_RECT || ...`, it would be better to define this expression as a function `isSelectToolType(ToolType)` in `src/core/control/ToolEnums.h`, and then call the function here. This would allow us to potentially add additional selection tools in the future without needing to change this part of the code.

File: src/core/control/Control.cpp
Comment (Technius) [inline src/core/control/Control.cpp:1790]: nit: `xournal->getTextEditor()->getText()` returns a pointer, so a move is unnecessary.
```suggestion
        Text* textobj = xournal->getTextEditor()->getText();
```

File: src/core/control/Control.cpp
Comment (Technius) [inline src/core/control/Control.cpp:1797]: Maybe I'm missing something, but under what scenario is `view` null? My understanding is that every page should have a view; if not, there is something wrong going on. Also, if `view` is indeed null, the `return` statement will prevent the tool switch from happening. 

I would replace this with an assertion:
```suggestion
        g_assert(view != nullptr);
```

File: src/core/control/Control.h
Comment (Technius) [inline src/core/control/Control.h:212]: nit: please remove this unnecessary line

**Review by Technius — COMMENTED** [review]
Thanks for the changes. I have a few nitpicks, but other than that, it looks good to me so far.

> But looking at the now occurring test-error in [Build_Test_Stage Test Xournal++ on MacOS](https://dev.azure.com/xournalpp/xournalpp/_build/results?buildId=21713&view=logs&jobId=69082843-5ef9-5486-682b-24fa9b8c0790) I'm not sure whether the issue is something I changed?

I think this is a temporary problem; you can ignore it.

**Review by Technius — APPROVED** [review]
LGTM. Merging in 24 hours if there are no objections.

File: src/core/control/Control.cpp
Comment (bhennion) [inline src/core/control/Control.cpp:1797]: Wait! Sorry I took a long time to check this out, but...
I fear there may be some "double ownership" cases here (textobj may be owned by both the layer and the selection). Let me check before merging.

Comment (bhennion) [reply src/core/control/Control.cpp:1797]: Ah, no, it seems to be ok. EditSelection(this->undoRedo, textobj, view, page); removes the element from the layer.

We need to lock the document's mutex though
```suggestion
        Text* textobj = xournal->getTextEditor()->getText();
        clearSelectionEndText();

        auto pageNr = getCurrentPageNo();
        XojPageView* view = xournal->getViewFor(pageNr);
        g_assert(view != nullptr);
        this->doc->lock()
        PageRef page = this->doc->getPage(pageNr);
        auto selection = new EditSelection(this->undoRedo, textobj, view, page);
        this->doc->unlock();
```

**Review by bhennion — APPROVED** [review]
LGTM. @Technius, you can merge at will :)

**Pull request merged**
